File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> res;
    void solve(vector<int>& candidates, int target, int ind, vector<int> &temp){
        if(target == 0){
            res.push_back(temp);
            return;
        }
        /*
        if(target < 0 || ind == candidates.size())
            return;
        
        // exclude the current element
        solve(candidates, target, ind + 1, temp);
        // include the current element
        temp.push_back(candidates[ind]);
        solve(candidates, target - candidates[ind], ind, temp);
        temp.pop_back();
        */
        for(int i = ind; i < candidates.size(); i++){
            // include the current element
            if(candidates[i] <= target){
            temp.push_back(candidates[i]);
            solve(candidates, target - candidates[i], i, temp);
            temp.pop_back();
            }
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> temp;
        solve(candidates, target, 0, temp);
        return res;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        solve(candidates, target, 0, new ArrayList<>(), ans);
        return ans;
    }

    public void solve(int[] candidates, int target, int start, List<Integer> temp, List<List<Integer>> ans){

        if(target < 0)
            return;
        if(target == 0){
            ans.add(new ArrayList<>(temp));
            return;
        }

        for(int i = start; i < candidates.length; i++){
            temp.add(candidates[i]);
            solve(candidates, target - candidates[i], i, temp, ans);
            temp.remove(temp.size() - 1);
        }
    }
}